lib/nbd.c

```diff
@@ -0,0 +1,792 @@
+                               bool dead_connection);
+
+const struct Curl_handler Curl_handler_nbd = {
+  "NBD",                                /* scheme */
+  ZERO_NULL,                            /* setup_connection */
+  ZERO_NULL,                            /* do_it */
+  ZERO_NULL,                            /* done */
+  ZERO_NULL,                            /* do_more */
+  nbd_connect,                          /* connect_it */
+  nbd_connection_state,                 /* connecting */
+  nbd_request_state,                    /* doing */
+  nbd_getsock,                          /* proto_getsock */
+  nbd_getsock,                          /* doing_getsock */
+  ZERO_NULL,                            /* domore_getsock */
+  ZERO_NULL,                            /* perform_getsock */
+  nbd_disconnect,                       /* disconnect */
+  ZERO_NULL,                            /* readwrite */
+  ZERO_NULL,                            /* connection_check */
+  PORT_NBD,                             /* defport */
+  CURLPROTO_NBD,                        /* protocol */
+  PROTOPT_NONE                          /* flags*/
+};
+
+/* Use for debugging builds. */
+#if defined(DEBUGBUILD) && !defined(CURL_DISABLE_VERBOSE_STRINGS)
+static const char *nbdstate_names[] = {
+  "STOP",
+  "EXPECTING_MAGIC",
+  "SENDING_CLIENT_FLAGS",
+  "SENDING_EXPORT_NAME",
+  "EXPECTING_EXPORT_DATA",
+  "EXPECTING_124_ZEROES",
+  "CONNECTED",
+  "SENDING_REQUEST",
+  "EXPECTING_REPLY",
+  "DOWNLOADING_DATA",
+};
+#endif
+
+static void debug_state(struct connectdata *conn, const char *what)
+{
+#if defined(DEBUGBUILD) && !defined(CURL_DISABLE_VERBOSE_STRINGS)
+  struct nbd_conn *nbdc = &conn->proto.nbdc;
+
+  infof(conn->data, "NBD %p %s state is %s\n",
+        nbdc, what, nbdstate_names[nbdc->state]);
+#endif
+}
+
+/* Only use this function for changing the state. */
+static void state(struct connectdata *conn, nbdstate newstate)
+{
+  struct nbd_conn *nbdc = &conn->proto.nbdc;
+#if defined(DEBUGBUILD) && !defined(CURL_DISABLE_VERBOSE_STRINGS)
+  if(nbdc->state != newstate)
+    infof(conn->data, "NBD %p state change from %s to %s\n",
+          nbdc, nbdstate_names[nbdc->state], nbdstate_names[newstate]);
+#endif
+  nbdc->state = newstate;
+}
+
+static CURLcode nbd_parse_url(struct connectdata *conn)
+{
+  CURLcode result;
+  struct nbd_conn *nbdc = &conn->proto.nbdc;
+
+  result = Curl_urldecode(conn->data, conn->data->state.path,
+                          0, &nbdc->exportname, NULL, TRUE);
+  if(result)
+    return result;
+
+  return CURLE_OK;
+}
+
+/* Create the protocol connection object. */
+static CURLcode nbd_connect(struct connectdata *conn, bool *done)
+{
+  struct nbd_conn *nbdc = &conn->proto.nbdc;
+  CURLcode result;
+  (void)done;
+
+  /* Initialize the connection structure. */
+  memset(nbdc, 0, sizeof(*nbdc));
+  state(conn, NBD_EXPECTING_MAGIC);
+  nbdc->recv_buf = malloc(NBD_MAX_MESSAGE_SIZE);
```

Comment:
Can't the ndb code use the existing read buffer libcurl already allocates and use for all other protocols? I realize smb.c uses its own buffer, but that was never a good design choice there and we should work on removing that. Unless there's a particular need, there's already an allocated buffer for reading remote resources into.